### PR TITLE
dataconnect: caching: add sqlite database logic

### PR DESCRIPTION
**Pull request by dconeybe**
This PR adds a sqlite database layer to `firebase-dataconnect` that manages cached query results and associated entities, complete with schema migration, secure database opening, and a suite of utility functions to streamline database interactions. 

### Highlights

* **New SQLite Database Logic**: Introduced a SQLite database implementation for offline caching within Firebase Data Connect. This includes core classes for database management, schema migration, and utility functions for interacting with SQLite.
* **Thread-Safe Database Operations**: The `DataConnectCacheDatabase` class is designed to be thread-safe, utilizing Kotlin coroutines and a single-threaded dispatcher to manage write transactions, preventing `SQLITE_BUSY` errors and ensuring ordered execution.
* **Robust Schema Management**: A `DataConnectCacheDatabaseMigrator` is included to handle database initialization and schema migrations, ensuring data integrity and compatibility across different versions. It also manages the SQLite `application_id` and `user_version` pragmas.
* **Enhanced SQLite Utilities**: New extension functions for `SQLiteDatabase` (`SQLiteDatabaseExts`) and a `SQLiteStatementBuilder` simplify common database operations, improve SQL logging, and provide a more robust way to construct parameterized queries.
* **New Utility Classes**: Added `ImmutableByteArray` for safe handling of byte arrays and `SemanticVersion` for structured version comparison and encoding, along with `StringUtil` for various string manipulations.
* **Extensive Unit Testing**: Comprehensive unit tests have been added for all new SQLite components, including database migration, opening, core cache operations, and utility classes, utilizing Kotest property-based testing for thorough validation of edge cases and data consistency.

<details>
<summary><b>Changelog</b></summary>

* **CHANGELOG.md**
    * Updated changelog to reflect the addition of SQLite database logic for offline caching.
* **DataConnectCacheDatabase.kt**
    * Added `DataConnectCacheDatabase` class, the core component for managing cached query results and entities in SQLite.
    * Implemented `initialize()` and `close()` methods for lifecycle management of the database.
    * Introduced `getQueryResult()` and `insertQueryResult()` for retrieving and storing query data and associated entities.
    * Utilizes Kotlin coroutines for thread-safe transaction management.
* **DataConnectCacheDatabaseMigrator.kt**
    * Added `DataConnectCacheDatabaseMigrator` for handling database schema creation and versioning.
    * Manages SQLite `application_id` and `user_version` pragmas to ensure database integrity and compatibility.
* **DataConnectSQLiteDatabaseOpener.kt**
    * Added `DataConnectSQLiteDatabaseOpener` to configure and open SQLite databases.
    * Ensures WAL journal mode, foreign key enforcement, full synchronous mode, and cell size checking for robustness.
* **SQLiteDatabaseExts.kt**
    * Added extension functions for `SQLiteDatabase` to provide simplified `execSQL` and `rawQuery` methods.
    * Includes enhanced logging for SQL statements and bind arguments.
    * Provides utilities for getting and setting the SQLite `application_id` and `last_insert_rowid`.
* **SQLiteStatementBuilder.kt**
    * Added `SQLiteStatementBuilder` for constructing SQL statements with automatic and re-usable binding numbering.
* **ImmutableByteArray.kt**
    * Added `ImmutableByteArray` as a wrapper for `ByteArray` to enforce immutability and provide safe handling.
* **SemanticVersion.kt**
    * Added `SemanticVersion` data class for representing and comparing semantic versions, including encoding to and decoding from integers.
* **StringUtil.kt**
    * Added `StringUtil` object with utility functions for hexadecimal string conversion and middle ellipsizing.
* **DataConnectCacheDatabaseMigratorUnitTest.kt**
    * Added unit tests for `DataConnectCacheDatabaseMigrator`, verifying correct application ID and schema migration behavior.
* **DataConnectCacheDatabaseUnitTest.kt**
    * Added unit tests for `DataConnectCacheDatabase`, covering initialization, closing, and the insertion and retrieval of query results with and without entities.
* **DataConnectSQLiteDatabaseOpenerUnitTest.kt**
    * Added unit tests for `DataConnectSQLiteDatabaseOpener`, ensuring correct database opening, file creation, and configuration settings.
* **QueryResultArb.kt**
    * Modified `QueryResultArb` to include `EntityRepeatPolicy` for more flexible generation of test data, allowing for repeated or mutated entities across samples.
* **SQLiteArbs.kt**
    * Added `SQLiteArbs` to provide Kotest arbitrary data generators for various SQLite column values, including nulls, integers, floats, doubles, strings, and byte arrays, with edge case handling.
* **SQLiteArbsUnitTest.kt**
    * Added unit tests for `SQLiteArbs` to validate the generation of diverse SQLite column values and their edge cases.
* **SQLiteDatabaseExtsUnitTest.kt**
    * Added unit tests for `SQLiteDatabaseExts`, covering SQL execution, query logging, and application ID management, ensuring correct behavior and logging output.
* **testutil/property/arbitrary/arbs.kt**
    * Added `semanticVersion` arbitrary generator for Kotest property testing.
* **ImmutableByteArrayUnitTest.kt**
    * Added unit tests for `ImmutableByteArray`, verifying its immutability, equality, and utility methods.
* **SemanticVersionUnitTest.kt**
    * Added unit tests for `SemanticVersion`, covering its encoding, decoding, and comparison logic.
* **StringUtilUnitTest.kt**
    * Added unit tests for `StringUtil`, verifying hexadecimal conversion for integers and byte arrays, and string ellipsizing functionality.
* **SqliteDatabaseRule.kt**
    * Added `SQLiteDatabaseRule`, a JUnit test rule for creating and managing SQLite databases (in-memory or file-based) for testing purposes.
* **StringCaseInsensitiveEquality.kt**
    * Added `StringCaseInsensitiveEquality`, a Kotest `Equality` implementation for case-insensitive string comparisons in tests.
* **TestUtils.kt**
    * Added utility functions `containLoneSurrogate()` and `shouldNotContainLoneSurrogates()` for verifying Unicode string integrity in tests.
* **ListNoRepeats.kt**
    * Added `listNoRepeat` arbitrary generator for Kotest, ensuring generated lists contain no repeated elements.
* **OffsetLengthOutOfRangeArb.kt**
    * Added `OffsetLengthOutOfRangeArb` for generating arbitrary offset and length values that are intentionally out of bounds for array access, useful for testing error handling.
* **OffsetLengthOutOfRangeArbUnitTest.kt**
    * Added unit tests for `OffsetLengthOutOfRangeArb` to verify its ability to generate out-of-range offset and length values.
</details>

### CODE REVIEW

File: firebase-dataconnect/src/main/kotlin/com/google/firebase/dataconnect/sqlite/DataConnectCacheDatabase.kt [hunk 1]
```diff
@@ -0,0 +1,559 @@
...
+          EntityParseFailureAction.Error,
+        )
+
+      val rehydrateResult = runCatching {
+        rehydrateQueryResult(queryResultProto, entityStructByEntityId)
+      }
+      rehydrateResult.onFailure {
+        logger.warn {
+          "rehydrateQueryResult failed failed for " +
+            "id=${sqliteQueryId.sqliteRowId}, queryId=${queryId.to0xHexString()} [knpe3t4f5b]"
```
[the remark this answers is not in the document]
Comment (dconeybe) [reply firebase-dataconnect/src/main/kotlin/com/google/firebase/dataconnect/sqlite/DataConnectCacheDatabase.kt]: Done.

File: firebase-dataconnect/src/main/kotlin/com/google/firebase/dataconnect/sqlite/DataConnectCacheDatabaseMigrator.kt [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    try {
+      sqliteDatabase.beginTransaction()
+
+      val userVersion = sqliteDatabase.version
+      if (userVersion == 0) {
+        logger.debug { "user_version is 0; initializing database" }
+        initializeDatabase()
+        val newUserVersion = SemanticVersion(1, 0, 0)
+        val newUserVersionInt = newUserVersion.encodeToInt()
+        logger.debug { "setting user_version to newUserVersionInt ($newUserVersion)" }
```
[the remark this answers is not in the document]
Comment (dconeybe) [reply firebase-dataconnect/src/main/kotlin/com/google/firebase/dataconnect/sqlite/DataConnectCacheDatabaseMigrator.kt]: Done
